File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        if(k <= 1){
            return 0;
        }
        
        int n = nums.size();
        
        int totalCount = 0;
        int i = 0, j = 0;
        long prod = 1;
        
        while(j < n){
            prod *= nums[j];
            
            while(prod >= k){
                prod /= nums[i++];
            }
            
            totalCount += (j - i) + 1;
            
            j++;
        }
        
        return totalCount;
    }
};

File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int n = nums.length;
        int count = 0;
        int i = 0, j = 0;
        long currProd = 1;
        while(j < n){
            currProd *= nums[j];
            while(i <= j && currProd >= k){
                currProd /= nums[i++];
            }
            count += j - i + 1;
            j++;
        }
        return count;
    }
}